src/rational/mat_q/cholesky_decomp.rs

```diff
@@ -0,0 +1,151 @@
+    ///
+    /// Returns the Cholesky decomposition of a Hermitian positive-definite matrix.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::rational::MatQ;
+    /// use std::str::FromStr;
+    ///
+    /// let matrix = MatQ::from_str("[[4, 12, -16],[12,37,-43],[-16,-43,98]]").unwrap();
+    ///
+    /// let l = matrix.cholesky_decomposition();
+    /// assert_eq!(matrix, &l * l.transpose());
+    /// ```
+    ///
+    /// # Panics ...
+    /// - if `self` is not a symmetric matrix.
+    ///
+    /// # Note
+    /// - the function may behave unintended if the provided matrix is not a
+    ///     Hermitian positive-definite matrix.
+    pub fn cholesky_decomposition(&self) -> MatQ {
+        assert!(self.is_symmetric(), "The provided matrix is not symmetric.");
+        // TODO: replace manual implementation with faster implementation from
+        // FLINT directly, once that is accessible through the FFI
+        let n = self.get_num_columns();
+
+        let mut a = self.clone();
+        let mut l = MatQ::identity(n, n);
+
+        for i in 0..n {
+            // get first entry and ith column
+            let a_ii = a.get_entry(0, 0).unwrap();
+            assert!(a_ii > Q::ZERO, "The matrix is not positive-definite.");
+            let column_a_i = match i {
+                0 => a.get_column(0).unwrap(),
+                _ => MatQ::new(i, 1)
+                    .get_column(0)
+                    .unwrap()
+                    .concat_vertical(&a.get_column(0).unwrap())
+                    .unwrap(),
+            } * (1 / (a_ii.sqrt()));
+
+            // produce L matrix recursively
+            let mut l_i = MatQ::identity(n, n);
+            l_i.set_column(i, &column_a_i, 0).unwrap();
+            l = l * l_i;
+
+            // update matrix A recursively
+            if i < n - 1 {
+                let b = a.get_submatrix(1, n - i - 1, 1, n - i - 1).unwrap();
+                let b_minus = (1 / a_ii)
+                    * a.get_submatrix(1, n - i - 1, 0, 0).unwrap()
+                    * a.get_submatrix(0, 0, 1, n - i - 1).unwrap();
+                a = b - b_minus;
+            }
+        }
+        l
+    }
+}
+
+#[cfg(test)]
+mod tes_cholesky_decomposition {
+    use crate::{
+        rational::{MatQ, Q},
+        traits::SetEntry,
+    };
+    use std::str::FromStr;
+
+    /// Ensure that a basic example (from Wikipedia) works.
+    #[test]
+    fn valid_input() {
+        let matrix = MatQ::from_str("[[4, 12, -16],[12,37,-43],[-16,-43,98]]").unwrap();
+
+        let l = MatQ::from_str("[[2, 0, 0],[6, 1, 0],[-8, 5, 3]]").unwrap();
+        assert_eq!(l, matrix.cholesky_decomposition());
+    }
+
+    /// Ensure that the function panics if a non-square matrix is provided
+    #[test]
+    #[should_panic]
+    fn non_square() {
+        let matrix = MatQ::new(3, 2);
+
+        matrix.cholesky_decomposition();
+    }
+
+    /// Ensure that the function panics if a non-positive-definite matrix is provided
```

Comment:
Does not match documentation, since that states that function does not check for pos. def. Function only checks for diag>0

Reply:
yes, somewhow overlooked that our sqrt panicks if provided with negative values. Fixed the documentation that the function captures this